vanilla/src/launch/java/org/spongepowered/vanilla/launch/plugin/resolver/DependencyResolver.java

```diff
@@ -0,0 +1,222 @@
+/*
+ * This file is part of Sponge, licensed under the MIT License (MIT).
+ *
+ * Copyright (c) SpongePowered <https://www.spongepowered.org>
+ * Copyright (c) contributors
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.vanilla.launch.plugin.resolver;
+
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.spongepowered.api.util.Tuple;
+import org.spongepowered.plugin.PluginCandidate;
+import org.spongepowered.plugin.PluginResource;
+import org.spongepowered.plugin.metadata.PluginDependency;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedHashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public final class DependencyResolver {
```

Comment:
I wonder if this would look cleaner overall as an instance on `VanillaPluginManager`? I try to avoid statics if I can.

Reply:
I was on the edge about doing that - don't really have a preference. I do tend to try to avoid statics from a testing point of view, but it makes little to no difference here because it's written in a way that doesn't store a state (rather, it's passed around). If you were going to make the plugin manager testable, on the other hand, then there'd be something to gain, but I suspect that might not happen.

As for your cleaner argument, I would disagree - creating an instance and storing it as opposed to a direct call isn't really cleaner, it's less so if anything. That's most certainly personal preference though. Static methods do have their place.

Reply:
Fair enough, your counter arguments are difficult to rebut.